### ISSUE TITLE
*Instance types should inherit from some basic Instance type which has "address"

### ISSUE THREAD

**Issue by roderik** [ethereum-ts/TypeChain#114]
We have in our contract library a GateKeeper contract which manages permissions for all contracts. 

We have a typescript function to create permissions which looks like this currently:

```
export async function createPermission<ContractInstance>(
  gateKeeper: GateKeeperContract,
  securedContractInstance: ContractInstance,
  permissionName: string,
  permissionManagerAddress: string,
  permissionRecipientAddress: string
) {
  const deployedGateKeeper = await gateKeeper.deployed();
  const role: string = await securedContractInstance[permissionName].call();

  await deployedGateKeeper.createPermission(
    permissionRecipientAddress,
    (securedContractInstance as any).address,
    role,
    permissionManagerAddress
  );
}
```

It would be much more clean if this could be:

```
export async function createPermission<ContractInstance extends Truffle.ContractInstance>(
  gateKeeper: GateKeeperContract,
  securedContractInstance: ContractInstance,
  permissionName: string,
  permissionManagerAddress: string,
  permissionRecipientAddress: string
) {
  const deployedGateKeeper = await gateKeeper.deployed();
  const role: string = await securedContractInstance[permissionName].call();

  await deployedGateKeeper.createPermission(
    permissionRecipientAddress,
    securedContractInstance.address,
    role,
    permissionManagerAddress
  );
}
```

Apart from the address field, it would also allow us to be a lot more flexible in generics.

**Comment by roderik** [comment]
FYI, the same goes for *Contract types, they also have an address field

**Comment by krzkaczor** [comment]
So, `*Instances` contracts lack address property. Yeah this is def bug on our side. I am gonna fix it asap. 

I am not sure about `*Contracts` having address field... arent they just to deploy instances? so they should not have address. Calling it in our tests gives us garbage like this:
```
DumbContract.address
Error: DumbContract has no network configuration for its current network id (4447).
undefined
```

**Comment by roderik** [comment]
In Truffle, if you run `deployer.deploy(Contract, ...params)` the Contract (which equates to the artifact) gets an address. That means you can bypass .deployed() if you just need the address for a follow up migration. 

Not debating if this is "clean" :), it is just something we use that works with Truffle.

**Comment by krzkaczor** [comment]
@roderik i get it. Okay, all your issues should be resolved now and released as `+ typechain@0.3.8`, also you will need to update `+ truffle-typings@1.0.4` if you used it.
